### PR TITLE
[WIP] Filesystem extension - special folders

### PR DESCRIPTION
**Pull request by Wend1go**
This extension will provide access to special folders, like documents, pictures, executable path, userdata, temp and desktop.
(I might also later add the possibility to read a text file into a string variable)

This is currently not working. I get the following error:
```
Uncaught TypeError: Cannot read property 'getFolderPictures' of undefined
    at Object.gdjs.NewSceneCode.eventsList0xb0a88 (code0.js:21)
    at gdjs.RuntimeScene.gdjs.NewSceneCode.func [as _eventsFunction] (code0.js:36)
    at gdjs.RuntimeScene.renderAndStep (runtimescene.js:232)
    at gdjs.SceneStack.step (scenestack.js:30)
    at runtimegame.js:354
    at gameLoop (runtimegame-pixi-renderer.js:375)
```
Probably easy to fix, will look into it tomorrow.

### CODE REVIEW

File: Extensions/FileSystem/filesystemtools.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * Get the path to 'Desktop' folder.
+ * @return {string} The path to the desktop folder
+ */
+gdjs.filesystem.getFolderDesktop = function (runtimeScene) {
+  const electron = runtimeScene.getGame().getRenderer().getElectron();
+
+  if (electron) {
+    return electron.getPath('desktop') ? electron.getPath('desktop') : '';
```
Comment (4ian) [inline Extensions/FileSystem/filesystemtools.js]: Could we name all functions `getDesktopPath` (i.e: `getXXXPath`)? Sounds more consistent to me with the electron `getPath` naming. "Folder" is mostly a metaphor for end users, technically prefer using path/filename.

You can write `return electron.getPath('desktop') || '';` (i.e: `a || b` instead of `a ? a : b`).

File: Extensions/FileSystem/filesystemtools.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  if (electron) {
+    return electron.getPath('exe') ? electron.getPath('exe') : '';
+  } else {
+    return '';
+  }
+}
+
+/**
+ * Get the path to 'userdata' folder.
+ * @return {string} The path to userdata folder
```
Comment (4ian) [inline Extensions/FileSystem/filesystemtools.js]: Thanks for the typing. You can also add everywhere:
`@param {gdjs.RuntimeScene} runtimeScene The current scene`.

before the @return to type the parameter and get autocompletion :)

File: Extensions/FileSystem/JsExtension.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      t(
+        "Access the filesystem of the operating system."
+      ),
+      "Matthias Meike",
+      "Open source (MIT License)"
+    ).setExtensionHelpPath("/all-features/filesystem");
+
+    extension
+      .addStrExpression(
+        "SpecialFolderDesktop",
```
Comment (4ian) [inline Extensions/FileSystem/JsExtension.js]: I think we can name them simply `DesktopPath`. I think Special does not have any value here (it's already "special" int the sense that we're giving to the user a method specially for it ;)) and as mentionned in another comment, I think Path would be more consistent than Folder with other programming languages/libraries.

(Expressions will be prefixed by the extension name, so we'll have `FileSystem::DesktopPath()` which looks good - or `FileSystem::GetDesktopPath()` if you prefer - but as there is no parameter we can think of DesktopPath almost like a constant so the "Get" could be skipped from the name).

**Comment by Wend1go** [comment]
Something seems to be wrong with updating via `node ./import-GDJS-Runtime.js`.
I initially named the extension "SpecialFolders" and placed it in a folder of the same name. Later I decided to rename it to "FileSystem". But upon running import-GDJS-Runitme the script tried to copy the files into the existing "SpecialFolders" folder which GD obviously could not find any more. I manually deleted it from `GDevelop/newIDE/app/resources/GDJS/Runtime/Extensions` and ran import-GDJS-Runtime again and it created the correct "FileSystem" folder with the correct files in it. But upon running a testgame, it still shows the same error.
Is there another place where I need to delete the old folder?

**Comment by 4ian** [comment]
> But upon running import-GDJS-Runitme the script tried to copy the files into the existing "SpecialFolders" folder which GD obviously could not find any more

What make you think that? `import-GDJS-Runtime.js` is just running calling `CopyRuntimeToGD.sh` or `CopyRuntimeToGD.bat` which are themselves just doing a dumb copy of the Extensions folder.

Can you show me the content of `newIDE/app/resources/GDJS/Runtime/Extensions`? This is where your extensions should be copied, so you should have your `FileSystem` folder here.

**Comment by 4ian** [comment]
The other hypothesis is that the include file "filesystemtools.js" is not picked correctly during export 🤔

**Comment by 4ian** [comment]
Can you also show me the output of the console after clicking on preview? (like the second screenshot [here](https://github.com/4ian/GDevelop/pull/828#issuecomment-450040615)).
I want to check if it's an error when copying the file (in which case we just need to find where/why the file is missing) or if there is no error (in which case, the js file is not even tried to be copied during export)

**Comment by Wend1go** [comment]
> What make you think that? 

The folder that I deleted contained the JsExtension.js + specialfoldertools.js and filesystemtools.js

> Can you show me the content of `newIDE/app/resources/GDJS/Runtime/Extensions`? This is where your extensions should be copied, so you should have your `FileSystem` folder here.

After deleting the SpecialFolders folder and rerunning the import-GDJS-Runtime.js it only contains the JsExtension.js and filesystemtools.js (with the proper content)

The extension seems to be loaded without errors in the project:
grafik

But the running game doesn't list the FileSystem extension folder:
grafik

**Comment by 4ian** [comment]
> rerunning the import-GDJS-Runtime.js it only contains the JsExtension.js and filesystemtools.js (with the proper content)

So the script is fine.

Can you show me the debugger console content after launching the preview? (not the console of the game, the console of GDevelop, the one showing "Loaded 8 JS extensions" :))

**Comment by 4ian** [comment]
To be sure it's not related to a bug in GD when previewing/exporting, I've created a blank game with only an action that uses an expression of an extension (DeviceSensors). Things seems to work properly, I can see the extension file and no errors related to it in both consoles.

image

**Comment by Wend1go** [comment]
Here is the console output after running the game. (just a text object with the pictures path assigned on key press)
grafik

**Comment by 4ian** [comment]
Really strange, let me check out your branch locally and try.

**Comment by blurymind** [comment]
There was a method to normalize paths, but it's inside the path module of nodejs.
https://millermedeiros.github.io/mdoc/examples/node_api/doc/path.html#path.normalize

It might not be a problem on windows, but its worth a try :)

**Comment by Wend1go** [comment]
Thanks for the link, this could be an option in case Windows doesn't understand the mixture of "/" and "\\".
I could even wrap those functions into the extension. `path.join` and `path.exists` look pretty useful too.

**Comment by 4ian** [comment]
Exposing these methods would be interesting, though I'm 99% sure that there won't be any issue using "/" (GDevelop itself even have some path handling methods coded using "/" as it's compatible with all OSes). `path.join` need the concept of arrays, which don't exist per se in GDevelop variables (though you can use structures), but you can expose an expression returning the path delimiter of the OS.

**Comment by Wend1go** [comment]
I added a string expression to get the path delimiter, so the problem should be solved now.

Concerning path.join, one could create an action that takes PATH1 and PATH2 as parameters and writes it into a scene variable. But I think this isn't needed any more now that we have the path delimiter.

**Comment by Wend1go** [comment]
So the last thing to add before this is ready to merge would be the icons.
I wonder if they are even displayed though.

**Comment by Wend1go** [comment]
The more functions I add, the more more I come to think it would be better to join it with the storage extension. These two become more and more similar.

**Comment by 4ian** [comment]
> I wonder if they are even displayed though. Are they needed for string expressions?

They are not displayed for now, but could be later so I would be good to have a least a simple icon for it :)

> The more functions I add, the more more I come to think it would be better to join it with the storage extension. These two become more and more similar.

Yes they are familiar. Storage is still something slightly more like an easy way to store data without worrying too much about where/how it's stored - the extensions is responsible for this.
While the FileSystem extension is offering more "low level" related functions (i.e: access to path, maybe later reading/writing files but directly without any "structure": the user is responsible for doing whatever we want to do).

So I think it's fine to keep them separated. Storage = High level, recommended way of storing user progress. FileSytem = low level, actions/conditions/expressions to deal with files and paths for certain use cases.
What do you think?

File: Extensions/FileSystem/JsExtension.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        "JsPlatform/Extensions/take_screenshot24.png",
+        "JsPlatform/Extensions/take_screenshot32.png"
+      )
+      .addCodeOnlyParameter('currentScene', '')
+      .addParameter("string", t("Directory"), "", false)
+      .getCodeExtraInformation()
+      .setIncludeFile(
+        "Extensions/FileSystem/filesystemtools.js"
+      )
+      .setFunctionName("gdjs.filesystem.getMakeDirectory");
```
Comment (4ian) [inline Extensions/FileSystem/JsExtension.js]: `gdjs.filesystem.makeDirectory`

File: Extensions/FileSystem/filesystemtools.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+/**
+ * Get the path delimiter specific to the operating system.
+ * @param {gdjs.RuntimeScene} runtimeScene The current scene
+ * @return {string} The path delimiter
+ */
+gdjs.filesystem.getPathDelimiter = function (runtimeScene) {
+  const electron = runtimeScene.getGame().getRenderer().getElectron();
+  const path = require('path');
```
Comment (4ian) [inline Extensions/FileSystem/filesystemtools.js]: This will probably fail in a browser where require is not defined, crashing the game.
Instead do:
`const path = typeof require !== 'undefined' ? require('path') : null;`

File: Extensions/FileSystem/filesystemtools.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+  if (electron && path) {
```
Comment (4ian) [inline Extensions/FileSystem/filesystemtools.js]: You don't need `electron` in this function.

File: Extensions/FileSystem/filesystemtools.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  }
+}
+
+/**
+ * Create a new directory at the given path.
+ * @param {gdjs.RuntimeScene} runtimeScene The current scene
+ * @param {string} directory The path to create a new directory
+ */
+gdjs.filesystem.getMakeDirectory = function (runtimeScene, directory) {
+  const electron = runtimeScene.getGame().getRenderer().getElectron();
```
Comment (4ian) [inline Extensions/FileSystem/filesystemtools.js]: You don't need electron in this function too.

File: Extensions/FileSystem/filesystemtools.js [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+  if (electron) {
+    const filesystem = require('fs');
```
Comment (4ian) [inline Extensions/FileSystem/filesystemtools.js]: Same here, you still need to protect against "require" being not defined.
require is only available when running in Node.js (which is the case in Electron, but not in browsers)

File: Extensions/FileSystem/filesystemtools.js [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+/**
+ * Check if the file or directory exists.
+ * @param {gdjs.RuntimeScene} runtimeScene The current scene
+ * @param {string} directory The path to the file or directory
+ * @return {boolean} True if fhe file or directory exists
+ */
+gdjs.filesystem.pathExists = function (runtimeScene, path) {
+  const electron = runtimeScene.getGame().getRenderer().getElectron();
```
Comment (4ian) [inline Extensions/FileSystem/filesystemtools.js]: Same here: remove electron, add a protection against missing require.

File: Extensions/FileSystem/filesystemtools.js [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+  
+  if (electron) {
+    const filesystem = require('fs');
+    return filesystem.existsSync(path) ? 1 : 0;
```
Comment (4ian) [inline Extensions/FileSystem/filesystemtools.js]: return `true` or `false` (i.e: just `return fs.existsSync(path)`). JavaScript may be forgiving about types, but we must not be, otherwise you'll get at some point very strange error or unexpected behavior that are hard to debug because you get a number instead of a boolean.

File: Extensions/FileSystem/filesystemtools.js [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+  } else {
+    return 0;
```
Comment (4ian) [inline Extensions/FileSystem/filesystemtools.js]: Same, always return the proper type: `return false;`.

**Comment by 4ian** [comment]
A few words about electron, Node.js and require, because I think there might be some confusions.

* Browsers have JavaScript engines. They don't provide `require` function, no access to filesystem. JavaScript engines still offer a few nices builtin APIs, for example `localStorage`.
* Node.js is a program embedding a JavaScript engine (V8, the Chrome engine) AND offering a function called `require` which allows to load "modules" AND offering "builtin" modules, including one called "fs".
* Electron is a program embedding Node.js and Chromium (the Chrome underlying rendering engine), allowing to do everything that you can do with Node.js AND offering other "builtin" modules. These modules are essentially useful to create apps. All are accessible from the module called `electron`.

It's important to know that because, we use them in this order in the game engine:
* We always try to stick to "standard" JavaScript without external APIs or modules. Meaning that games won't have access to file system (which is good, as this would open dangerous security holes to website otherwise). We still are able to implement most of what is needed, for example an easy to use "Storage" extension that uses `localStorage`.
* IF we want to expose operating system things, we can't do it unless we're running on Node.js. This is almost never the case for GDevelop games, because the game engine is made to run in a browser and Node.js is not a browser, just a JavaScript engines.
* BUT electron is now offering a browser + Node.js, exposing also the "require" function. So most of the time, if we are running with Node.js, this is because we are running in Electron (so we both have `electron` and `require`).

These things might differ in the future though. (In short, Electron has been criticized for offering too much opportunities for security issues, because it might be used to display untrusted contents (i.e: random websites) AND giving them an access to Node.js APIs (i.e: using `require`), which might allow to do very bad things. Fortunately this is not the case with GDevelop, nor GDevelop games, because GDevelop is a "self contained" app, meaning that all that we're executing is trusted and any link is opening in a real browser with no Node.js access (and same for GDevelop games published with Electron)).
So we must use ensure that our code is resilient against missing things, mainly:

* whenever you're using `require`, you're assuming that you're running in Node.js (or something else that provides you with a require function) (because yes things can't be that simple ;) Bundlers like "Webpack" allow you to use require and export a single JavaScript file - anyway it's out of our scope here). So we must check for the existence of `require` (`typeof require !== 'undefined'`)
* whenever you're using `electron`, you're assuming electron existence and must check for its existence (`if (electron) {...}`)
* and if you're using only require, don't check for electron, otherwise we're restricting some part of the extension to be running in Electron WHILE later we can imagine someone wanting for some reason to run a "game" only in Node.js. Chances are small but let's try to keep things right from the beginning.

tl;dr: I'm being pedantic, let's make sure to only use electron when needed :D (I've left comments where things are to be adapted ;))

**Comment by Wend1go** [comment]
You are not pedantic, I really appreciate your detailed explanations. These really help me understand the way GD works and make me write better code, so that you'll have less work reviewing in the future.

Basically I agree with you about the high and low level functions in `Storage` and `FileSystem`. I was just concerned about the `Delete a file` function in `Storage`which is a low level function in the high level extension. I was just about to add it in `FileSystem` until I noticed that it already exists.

**Comment by 4ian** [comment]
> Basically I agree with you about the high and low level functions in Storage and FileSystem. I was just concerned about the Delete a file function in Storage which is a low level function in the high level extension. I was just about to add it in FileSystem until I noticed that it already exists.

Here I think the issue is that the Storage extension action description and name is not adapted anymore to what it's doing. Before, for native games, this action was effectively writing "real files", so an action for deleting a file made sense. Now, as all actions and conditions of Storage are acting on some storage in the browser (localStorage), it would make sense to rename this "Delete stored data" so that it's clear there are no real files involved. I'll do this!

**Comment by 4ian** [comment]
Looks good now :) Are you able to create an icon for it? I would rather avoid keeping the screenshot here - because at some point I'd like to show the icons in the list of actions/conditions/expressions :)

**Comment by Wend1go** [comment]
Here are the icons, should be ready to merge now.

**Comment by 4ian** [comment]
Sounds good, I'll merge this. Tests were failing also on master after changes in instructions and tests not being updated.

**Comment by 4ian** [comment]
Thanks for working on this :) 
I think for screenshot we can prepare the wiki page and have an example or at least a screenshot (;)) of the action used with FileSystem in the wiki.

**Comment by blurymind** [comment]
thank you @Wend1go ! :)
